src/astar.cpp

```diff
@@ -508,59 +471,66 @@ static ASR_RETVAL fpathTracePath(const PathfindContext& context, PathCoord src,
 	return retval;
 }
 
-#if DEBUG
-extern uint32_t selectedPlayer;
-#endif
-
-ASR_RETVAL fpathAStarRoute(MOVE_CONTROL *psMove, PATHJOB *psJob)
+ASR_RETVAL fpathAStarRoute(std::list<PathfindContext>& fpathContexts,
+	MOVE_CONTROL *psMove, PATHJOB *psJob)
 {
-	ASR_RETVAL      retval = ASR_OK;
+	ASSERT_OR_RETURN(ASR_FAILED, psMove, "Null psMove");
+	ASSERT_OR_RETURN(ASR_FAILED, psJob, "Null psMove");
+
+	ASR_RETVAL retval = ASR_OK;
 
-	bool            mustReverse = true;
+	bool mustReverse = false;
 
-	const PathCoord tileOrig(map_coord(psJob->origX), map_coord(psJob->origY));
-	const PathCoord tileDest(map_coord(psJob->destX), map_coord(psJob->destY));
-	const PathNonblockingArea dstIgnore(psJob->dstStructure);
+	const PathCoord tileOrig = psJob->blockingMap->worldToMap(psJob->origX, psJob->origY);
+	const PathCoord tileDest = psJob->blockingMap->worldToMap(psJob->destX, psJob->destY);
 
-#if DEBUG
-	const DROID *psDroid = IdToDroid(psJob->droidID, selectedPlayer);
-	const bool isDroidSelected = (psDroid && psDroid->selected);
-#endif
+	if (psJob->blockingMap->isBlocked(tileOrig.x, tileOrig.y)) {
+		debug(LOG_NEVER, "Initial tile blocked (%d;%d)", tileOrig.x, tileOrig.y);
+	}
+	if (psJob->blockingMap->isBlocked(tileDest.x, tileDest.y)) {
+		debug(LOG_NEVER, "Destination tile blocked (%d;%d)", tileDest.x, tileDest.y);
+	}
+
+	const PathNonblockingArea dstIgnore(psJob->dstStructure);
 
-	NearestSearchPredicate pred(tileDest);
+	NearestSearchPredicate pred(tileOrig);
 
-	PathCoord endCoord;  // Either nearest coord (mustReverse = true) or orig (mustReverse = false).
+	PathCoord endCoord;
 
+	// Caching reverse searches.
 	std::list<PathfindContext>::iterator contextIterator = fpathContexts.begin();
 	for (; contextIterator != fpathContexts.end(); ++contextIterator)
 	{
 		PathfindContext& pfContext = *contextIterator;
-		if (!pfContext.matches(psJob->blockingMap, tileDest, dstIgnore))
+		if (!pfContext.matches(psJob->blockingMap, tileDest, dstIgnore, /*reverse*/true))
 		{
 			// This context is not for the same droid type and same destination.
 			continue;
 		}
 
+		const PathExploredTile& pt = pfContext.tile(tileOrig);
 		// We have tried going to tileDest before.
-		if (pfContext.map[tileOrig.x + tileOrig.y * mapWidth].iteration == pfContext.iteration
-		    && pfContext.map[tileOrig.x + tileOrig.y * mapWidth].visited)
+		if (pfContext.isTileVisited(pt))
 		{
 			// Already know the path from orig to dest.
 			endCoord = tileOrig;
```

Comment:
```suggestion
			endCoord = tileOrig;
			break;
```

Probably missing a `break` here? (This seems to make it harder to trigger the endless lag case, but there must be another path because it's still possible.)